### PR TITLE
Add exhaustive testing to ValueRanges, fix bugs

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95063
* #94985
* #94944
* __->__ #94939
* #94906

Since I didn't want to deal with nondeterministic tests, I went the exhaustive testing route for a fixed list of constants to look at. The tests generate random ranges, propagate the range through the function, and then pick elements in the range and check that the result on the operation is in the resulting range. This caught bugs in log, sqrt and pow.

My resolution for pow was a little special, because I had trouble figuring out the correct semantics under all inputs domains. Instead, I picked two input domains (pow on two point ranges, and pow where exponent is known) and only implemented those. Everything else we give up. I think this is unlikely to affect perf.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

File: torch/utils/_sympy/value_ranges.py [hunk 1]
```diff
@@ -261,31 +261,36 @@ def exp(x):
...
+                return ValueRanges.unknown()
+            return ValueRanges.wrap(r)
+        elif b.lower == b.upper and b.lower >= 0:
+            i = ValueRanges.wrap(1)
+            for _ in range(b.lower):
+                i = cls.mul(i, a)
+            return i
+        else:
+            # This is fairly difficult to analyze, so give up for anything
+            # complicated
```
Comment (lezcano) [inline torch/utils/_sympy/value_ranges.py:292]: Right, I see we were missing a couple cases here, yep.

File: torch/utils/_sympy/value_ranges.py [hunk 2]
```diff
@@ -261,31 +261,36 @@ def exp(x):
 
     @staticmethod
     def log(x):
+        if x.lower <= 0:
+            return ValueRanges.unknown()
         return ValueRanges.increasing_map(
             x, lambda y: -sympy.oo if y <= 0 else sympy.log(y)
```
Comment (lezcano) [inline torch/utils/_sympy/value_ranges.py:267]: With the new case, we can now simply use `smpy.log`  here?

**Review by lezcano — APPROVED** [review]

File: torch/utils/_sympy/value_ranges.py (hunk 1 above)
Comment (lezcano) [reply torch/utils/_sympy/value_ranges.py:292]: For future reference, it's probably enough to analyse the signs of a and b separately and have those 4 cases, but yeah, probably not worth the effort.

Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:292]: PRs welcome! The problem I had was that you also have different regimes for a < 1 vs > 1 vs = 1. PAIN

File: test/test_value_ranges.py [hunk 3]
```diff
@@ -119,15 +122,45 @@ def ceil(x):
         return math.ceil(x)
 
 
+def valid_unary(fn, v):
```
Comment (eellison) [inline test/test_value_ranges.py:125]: we could also run the reference first and see if it throws, then skip

File: test/test_value_ranges.py [hunk 4]
```diff
@@ -153,6 +184,48 @@ def test_binary_ref(self, fn):
...
+
+    # This takes about 4s for all the variants
+    @parametrize("fn", BINARY_OPS)
+    def test_binary_ref_range(self, fn):
+        vals = [-sympy.oo, *LESS_CONSTANTS, sympy.oo]
+        for a, b in itertools.product(generate_range(vals), repeat=2):
+            # don't attempt pow on exponents that are too large (but oo is OK)
+            if fn == "pow" and b.upper > 4 and b.upper != sympy.oo:
+                continue
+            with self.subTest(a=a, b=b):
```
Comment (eellison) [inline test/test_value_ranges.py:216]: `subTest` 😮

File: torch/utils/_sympy/value_ranges.py [hunk 5]
```diff
@@ -261,31 +261,36 @@ def exp(x):
 
     @staticmethod
     def log(x):
+        if x.lower <= 0:
```
Comment (eellison) [inline torch/utils/_sympy/value_ranges.py:264]: if this is has free symbol it would throw. but we can handle that when we get here

**Review by eellison — APPROVED** [review]
nice !

File: test/test_value_ranges.py (hunk 3 above)
Comment (ezyang) [reply test/test_value_ranges.py:125]: Still good to avoid trying to eval something silly like 2 ** 10000 which sympy will do symbolically lol

**Review by nunoplopes — APPROVED** [review]
LGTM!

**Comment by ezyang** [comment]
It's not intentional. You won't be able to revert this PR, please submit a forward fix PR.

**Comment by yanbing-j** [comment]
Please review the forward fix PR https://github.com/pytorch/pytorch/pull/95688. Thanks!

**Pull request closed without merging**